### ISSUE TITLE
COBOL - COPY REPLACING Invalid Parsing Error Message

### ISSUE THREAD

**Issue by davidkcoblentz** [IBM/zopeneditor-about#181]
# Development environment used
Version: 1.61.2 (user setup)
Commit: 6cba118ac49a1b88332f312a8f67186f7f3c1643
Date: 2021-10-19T14:57:20.575Z
Electron: 13.5.1
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.17763

### IBM Z Open Editor extension:
V1.2.7

# Problem Description
Part of the COBOL COPY statement includes the REPLACING phrase. Z Open Editor is issuing an "Invalid parsing" error in the Copy Book source as compared to no error on the mainframe compiler. The COBOL source example below is from a very large legacy system that utilizes many COPY REPLACING statements in it's COBOL source.  The below COPY REPLACING statement uses a valid copybook statement using copybook member STAC02 with a REPLACING phrase. By hovering over the Copy Book name, Z Open Editor does show the copy source accurately but with a parsing error message. The source from the COBOL source program and STAC02 copybook are listed below.
### Example of Z Open Editor Parsing Error Message:
![image](https://user-images.githubusercontent.com/80063515/139470516-5a1d3d28-2520-45b5-bfef-1c079ab08840.png)

### COBOL source excerpt of the COPY REPLACING statement:
```
....5...10...15...20...25...30...35...40...45...50...55...60...65...70...75...80
       01  STAC02                          COPY STAC02 REPLACING  ==01
                                                                 STAC02
                                                                 == BY
           ==    ==.

```
### STAC02 Copy Book source:
```
       01  STAC02 SYNCHRONIZED.                                                 
         03  TABLE-ENTRY-LENGTH    PICTURE 9(5)         COMPUTATIONAL.          
         03  FLEET-TYPE            PICTURE 99.                                  
```
### Compile Listing from the IBM mainframe compiler:
```
 BLW=0   AC0 000696         01  STAC02                                                                STAC02                      6P
             000697                                             COPY STAC02 REPLACING  ==01
             000698                                                                   STAC02
             000699                                                                   == BY
             000700             ==    ==.
             000701C            SYNCHRONIZED.                                                00000010
 BLW=0   AC0 000702C          03  TABLE-ENTRY-LENGTH    PICTURE 9(5)         COMPUTATIONAL.  00000020             1       4       4
 BLW=0   AC4 000703C          03  FLEET-TYPE            PICTURE 99.                          00000030             5       6       2
 BLW=0   000 000704         01  FILLER     PIC X(100).                                                FILLER                    100
```

# Observed Behavior
Z Open Editor lists a parsing error message in Copybook STAC02 indicating some type of parsing issue "Copybook 'STAC02' contains the following error: "DataDescriptionEntryClause" is invalidparsing" when there is no error according to COBOL syntax rules.

# Expected Behavior
Z Open Editor to parse the COPY REPLACING statement without error indications.

**Comment by FALLAI-Denis** [comment]
Hi @davidkcoblentz 

The problem encountered does not necessarily come from the very particular context of use of the COPY directive as illustrated here on a minimal program which implements the same use case:

![image](https://user-images.githubusercontent.com/63861965/140029198-73a0d5c7-c640-4c99-966e-90133db0f400.png)

I think there must be some other issue that is triggering the one you are having.

We should have all of the COBOL source code and COPYs used to diagnose the problem.

For my part, I also encounter problems on uses of COPY and on the use of conditional compilation directives which totally disrupt the operation of the COBOL Language Server, at least rendering the Z Open Editor unusable (confer #154 issue).

Note: your compiler does not seem to be the most recent ... COBOL II?

**Comment by kkrein** [comment]
Ich bin derzeit nicht im Büro.
Während meiner Abwesenheit bitte bei Postkorb SCM-Admin melden!

Die Rückmeldung bezieht sich auf ein Mail mit folgendem Thema:
Re: [IBM/zopeneditor-about] COBOL - COPY REPLACING Invalid Parsing Error
Message (Issue #181)

____________________________________________________________________________________________
Gesendet (c) GRZ/RSG Linz 2021 Agent 'Abwesenheit'

----------------------------------------
Raiffeisen Software GmbH
Firmenbuchnummer: 86804d, Landesgericht Linz
UID: ATU 36961105

Der Austausch von Nachrichten mit dem oben angeführten Absender via E-Mail
dient ausschließlich Informationszwecken. Rechtsgeschäftliche Erklärungen
dürfen über dieses Medium nicht ausgetauscht werden.
Correspondence with above mentioned sender via e-mail is only for
information purposes. This medium is not to be used for the exchange of
legally-binding communications.
----------------------------------------

**Comment by gplambert** [comment]
@davidkcoblentz - thank you for using our editor and thank you for bringing this bug to our attention.  The issue is that the element ==01  STAC02== is spanning multiple lines which the editor is not parsing correctly.  Until we have a fix for this in place, a workaround to no longer see the error message would be to put that element on one line.

**Comment by FALLAI-Denis** [comment]
@gplambert 

Are you sure this is the problem?
Cf. [my test above](https://github.com/IBM/zopeneditor-about/issues/181#issuecomment-958745689): no problem for same use case...

For my part, I rather think that there are problems in the management of COPYBOOKs in Language Server and that at times, Language Server gets lost and starts to report errors wrongly.

**Comment by davidkcoblentz** [comment]
Thank you and yes that work around is an easy fix.  Thanks for your attention to this.  I am excited to start using VS Code for all of my editing and research needs and all of the hard work you have put into IBM zOS Open Editor and Zowe!  It is the best editor I have seen yet and the only one that includes almost all of the editing power and features of the ISPF/PDF Editor on the mainframe I look forward to each of your future releases. Thanks for giving me a great editor that I can use on my PC!

Thank you,
***@***.***

**Comment by davidkcoblentz** [comment]
We are currently on COBOL 4.2 but should be 6.2 within 2 or 3 months.

Please see attached COBOL source for the program and copy book STAC02.  There are aver 93 copy books ion the source so I am hoping you can re-create the issue with just these 2 source components. If you need more please let me know.

Thank you,
***@***.***

**Comment by davidkcoblentz** [comment]
[AME026C.txt](https://github.com/IBM/zopeneditor-about/files/7477503/AME026C.txt)
[STAC02.txt](https://github.com/IBM/zopeneditor-about/files/7477512/STAC02.txt)

**Comment by FALLAI-Denis** [comment]
Hi,

I think it is rather the absence of a period at the end of the declaration of level 01 that causes the problem in Language Server (even if it is correct at the compilation time).

This is the error message that appears when the COPYBOOK is not found:

![image](https://user-images.githubusercontent.com/63861965/140489255-3cd28705-5771-4a9f-b370-19b730a4bd7d.png)

When the COPYBOOK is found, the error message on point disappears, and an error message on the structure of the COPYBOOK appears:

![image](https://user-images.githubusercontent.com/63861965/140489585-1decbdc0-d5aa-4b8e-a159-40941ef01a6b.png)

If I add a dot on the declaration of level 01 (which will cause a compilation error), then there is no longer an error at the Language Server level, even if the COPY directive is indicated on the same line than level 01:

![image](https://user-images.githubusercontent.com/63861965/140490066-42a8637d-38a3-4f50-b434-5afa072f9535.png)

The code below should compile correctly (I have not tested) and in addition is compatible with the Language Server, but there are impacts on the source code...

![image](https://user-images.githubusercontent.com/63861965/140491342-9c4294a4-47af-4e01-b3ea-dcf296a183af.png)

In any case, this shows that the problem is not the presence of the COPY directive on the same line as the declaration of level 01.


Regarding the compiler, I was deceived by the fact that you are using the Compuware compilation utility, which strongly modifies the standard Listing of the IBM Enterprise COBOL compiler.

**Comment by davidkcoblentz** [comment]
Hi,

Thank you for the excellent analysis.

I did some additional searching of our source since I knew we had many COBOL source programs that had COPY REPLACING syntax along with the SYNCHRONIZED phrase on the 01 level.  I came across another example and there was no error message. I discovered the parsing error only appears if it can not find a match for the REPLACING string **as seen in the modified example source below**. If the REPLACING string is found, no error is given by zOS Open Editor regarding the copy book. But a parsing error **is** given if the string is not found even if the COPY statement is completed on the one line.  

So I have 2 questions,

**1. Should zOS Open Editor issue a parsing error if the REPLACING string is not found in the copy book?**
I wouldn't think so. I would think there are times when it is acceptable that the REPLACING string is not found. Maybe a warning instead (yellow) or even a setting that allows developers to choose if they want strict checking of REPLACING strings.
**2. It appears that zOS Open Editor is not deriving the same REPLACING string as the COBOL compiler when the delimiters (==) span multiple lines as are derived according to COBOL continuation rules?**  
The original zOS Open Editor error was indicating the string was NOT FOUND.  Maybe because Open Editor may include all of the blanks on the second line when creating the REPLACING string. I'm not sure how it forms the string as it spans continuation lines.

### Minimized source example showing different ways of coding the COPY directive.
![image](https://user-images.githubusercontent.com/80063515/140630317-f861389c-396c-4314-b80a-79a721cfeb0a.png)

I hope this helps clarify the issue.

**Comment by kkrein** [comment]
Ich bin derzeit nicht im Büro. Ich kehre zurück am  06.12.2021.
Während meiner Abwesenheit bitte bei Postkorb SCM-Admin melden!

Die Rückmeldung bezieht sich auf ein Mail mit folgendem Thema:
Re: [IBM/zopeneditor-about] COBOL - COPY REPLACING Invalid Parsing Error
Message (Issue #181)

____________________________________________________________________________________________
Gesendet (c) GRZ/RSG Linz 2021 Agent 'Abwesenheit'

----------------------------------------
Raiffeisen Software GmbH
Firmenbuchnummer: 86804d, Landesgericht Linz
UID: ATU 36961105

Der Austausch von Nachrichten mit dem oben angeführten Absender via E-Mail
dient ausschließlich Informationszwecken. Rechtsgeschäftliche Erklärungen
dürfen über dieses Medium nicht ausgetauscht werden.
Correspondence with above mentioned sender via e-mail is only for
information purposes. This medium is not to be used for the exchange of
legally-binding communications.
----------------------------------------

**Comment by davidkcoblentz** [comment]
Just for reference, I pasted the comparison and replacement rules from the COBOL 4.2 manual. Note item 6 where multiple spaces are reduced to a single space between text words before comparison.  This explains in our sample code how it treats spaces spanning lines inside of Pseudo-text delimiters (==). It is doing sequential word matching. zOS Open Editor must not be following these rules when preparing the Pseudo-text string for comparison.  Also note that since the comparison is in AREA A the replacing string (operand-2) is 4 blanks. This is to make sure the SYNCHRONIZED. word ends up positioned in AREA B of the resulting source after replacements are done.

### COBOL 4.2 Replacement and Comparison Rules
1. Arithmetic and logical operators are considered text words and can be
replaced only through the pseudo-text option.
2. Beginning and ending blanks are not included in the text comparison process.
Embedded blanks are used in the text comparison process to separate multiple
text words.
3. When operand-1 is a figurative constant, operand-1 matches only the same exact
figurative constant. For example, if ALL "AB" is specified in the library text,
then "ABAB" is not considered a match; only ALL "AB" is considered a match.
4. When replacing a PICTURE character-string, the pseudo-text option should be
used; to avoid ambiguities, pseudo-text-1 should specify the entire PICTURE
clause, including the keyword PICTURE or PIC.
5. Any separator comma, semicolon, or space preceding the leftmost word in the
library text is copied into the source text. Beginning with the leftmost library
text word and the first operand-1 specified in the REPLACING option, the
entire REPLACING operand that precedes the keyword BY is compared to an
equivalent number of contiguous library text words.
**6. operand-1 matches the library text if, and only if, the ordered sequence of text
words in operand-1 is equal, character for character, to the ordered sequence of
library words. For national characters, the sequence of national characters
must be equal, national character for national character, to the ordered
sequence of library words. For matching purposes, each occurrence of a
comma or semicolon separator and each sequence of one or more space
separators is considered to be a single space. However, when operand-1
consists solely of a separator comma or semicolon, it participates in the match
as a text word (in this case, the space following the comma or semicolon
separator can be omitted).
When the library text contains a closing quotation mark that is not
immediately followed by a separator space, comma, semicolon, or period, the
closing quotation mark is considered a separator quotation mark.**
7. If no match occurs, the comparison is repeated with each successive operand-1,
if specified, until either a match is found or there are no further REPLACING
operands.
8. Whenever a match occurs between operand-1 and the library text, the
associated operand-2 is copied into the source text.
9. The COPY statement with REPLACING phrase can be used to replace parts of
words. By inserting a dummy operand delimited by colons into the program
text, the compiler will replace the dummy operand with the desired text.
Example 3 shows how this is used with the dummy operand :TAG:.
The colons serve as separators and make TAG a stand-alone operand.
10. When all operands have been compared and no match is found, the leftmost
library text word is copied into the source text.
11. The next successive uncopied library text word is then considered to be the
leftmost text word, and the comparison process is repeated, beginning with
the first operand-1. The process continues until the rightmost library text word
has been compared.
12. Comment lines or blank lines occurring in the library text and in pseudo-text-1
are ignored for purposes of matching; and the sequence of text words in the
library text and in pseudo-text-1 is determined by the rules for reference
format. Comment lines or blank lines appearing in pseudo-text-2 are copied
into the resultant program unchanged whenever pseudo-text-2 is placed into
the source text as a result of text replacement. Comment lines or blank lines
appearing in library text are copied into the resultant source text unchanged
with the following exception: a comment line or blank line in library text is
not copied if that comment line or blank line appears within the sequence of
text words that match pseudo-text-1.
13. Text words, after replacement, are placed in the source text according to
Standard COBOL 85 format rules.
14. When text words are placed in the source text, additional spaces are
introduced only between text words where there already exists a space
(including the assumed space between source lines).
15. COPY REPLACING does not affect the EJECT, SKIP1, SKIP2, SKIP3, or TITLE
compiler-directing statements.

**Comment by phaumer** [comment]
Fixed in Z Open Editor v2.0.0
